File: apps/csharp/design-pattern/src/DesignPattern/Patterns/Singleton/Logger.cs
namespace DesignPattern.Patterns.Singleton;

public class SimpleLogger
{
    private readonly List<string> _messages = new();

    public void Log(string message) => _messages.Add(message);

    public IReadOnlyList<string> Messages => _messages.AsReadOnly();

    public string LatestMessage => _messages.Count > 0 ? _messages[^1] : "";

    public int MessageCount => _messages.Count;
}

public sealed class SingletonLogger
{
    private static readonly Lazy<SingletonLogger> _instance =
        new(() => new SingletonLogger());

    private readonly List<string> _messages = new();

    private SingletonLogger() { }

    public static SingletonLogger Instance => _instance.Value;

    public void Log(string message) => _messages.Add(message);

    public IReadOnlyList<string> Messages => _messages.AsReadOnly();

    public string LatestMessage => _messages.Count > 0 ? _messages[^1] : "";

    public int MessageCount => _messages.Count;

    // For testing only
    internal void Reset() => _messages.Clear();
}

File: apps/php/design-pattern/src/Singleton/Logger.php
<?php

declare(strict_types=1);

namespace DesignPattern\Singleton;

class SingletonLogger
{
    private static ?SingletonLogger $instance = null;

    /** @var string[] */
    private array $log = [];

    private function __construct()
    {
        // 外部からのインスタンス化を禁止
    }

    private function __clone()
    {
        // クローンを禁止
    }

    public static function getInstance(): self
    {
        if (self::$instance === null) {
            self::$instance = new self();
        }
        return self::$instance;
    }

    public static function resetInstance(): void
    {
        self::$instance = null;
    }

    public function log(string $message): void
    {
        $this->log[] = $message;
    }

    /** @return string[] */
    public function getLog(): array
    {
        return $this->log;
    }

    public function clear(): void
    {
        $this->log = [];
    }
}
